contracts/modules/elrond-wasm-module-governance/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+		if !self.proposal_exists(proposal_id) {
+			return GovernanceProposalStatus::None;
+		}
+
+		let queue_block = self.proposal_queue_block(proposal_id).get();
+		if queue_block > 0 {
+			return GovernanceProposalStatus::Queued;
+		}
+
+		let current_block = self.blockchain().get_block_nonce();
+		let proposal_block = self.proposal_start_block(proposal_id).get();
+		let voting_delay = self.voting_delay_in_blocks().get();
+		let voting_period = self.voting_period_in_blocks().get();
+
+		let voting_start = proposal_block + voting_delay;
+		let voting_end = voting_start + voting_period;
+
+		if current_block < voting_start {
+			return GovernanceProposalStatus::Pending;
+		}
+		if current_block >= voting_start && current_block < voting_end {
+			return GovernanceProposalStatus::Active;
+		}
+
+		let total_votes = self.total_votes(proposal_id).get();
+		let total_downvotes = self.total_downvotes(proposal_id).get();
+		let quorum = self.quorum().get();
+
+		if total_votes > total_downvotes && total_votes - total_downvotes >= quorum {
+			return GovernanceProposalStatus::Succeeded;
+		} else {
+			return GovernanceProposalStatus::Defeated;
+		}
+	}
+
+	#[view(getProposer)]
+	fn get_proposer(&self, proposal_id: usize) -> OptionalArg<Address> {
+		if !self.proposal_exists(proposal_id) {
+			OptionalArg::None
+		} else {
+			OptionalArg::Some(self.proposals().get(proposal_id).proposer)
+		}
+	}
+
+	#[view(getProposalDescription)]
+	fn get_proposal_description(&self, proposal_id: usize) -> OptionalArg<BoxedBytes> {
+		if !self.proposal_exists(proposal_id) {
+			OptionalArg::None
+		} else {
+			OptionalArg::Some(self.proposals().get(proposal_id).description)
+		}
+	}
+
+	#[view]
+	fn get_proposal_actions(
+		&self,
+		proposal_id: usize,
+	) -> MultiResultVec<GovernanceActionAsMultiArg<Self::BigUint>> {
+		if !self.proposal_exists(proposal_id) {
+			return Vec::new().into();
+		}
+
+		let actions = self.proposals().get(proposal_id).actions;
+		let mut actions_as_multiarg = Vec::with_capacity(actions.len());
+
+		for action in actions {
+			actions_as_multiarg.push(action.into_multiarg());
+		}
+
+		actions_as_multiarg.into()
+	}
+
+	// private
+
+	fn require_payment_token_governance_token(&self) -> SCResult<()> {
+		require!(
+			self.call_value().token() == self.governance_token_id().get(),
+			"Only Governance token accepted as payment"
+		);
+		Ok(())
+	}
+
+	fn require_valid_proposal_id(&self, proposal_id: usize) -> SCResult<()> {
+		require!(
+			self.is_valid_proposal_id(proposal_id),
+			"Invalid proposal ID"
+		);
+		Ok(())
+	}
+
+	fn is_valid_proposal_id(&self, proposal_id: usize) -> bool {
+		proposal_id >= 1 && proposal_id <= self.proposals().len()
+	}
+
+	fn proposal_exists(&self, proposal_id: usize) -> bool {
+		self.is_valid_proposal_id(proposal_id) && !self.proposals().item_is_empty(proposal_id)
+	}
+
+	fn total_gas_needed(&self, actions: &[GovernanceAction<Self::BigUint>]) -> u64 {
+		let mut total = 0;
```

Comment:
Check for overflow maybe?

Reply:
Added check for each action to have less than max gas limit per block (1.5 billion), so this can never overflow.